packages/api/src/routes/pagination.ts

```diff
@@ -0,0 +1,141 @@
+import { numericValue } from "@metriport/core/external/carequality/ihe-gateway-v2/schema";
+import { Config } from "@metriport/core/util/config";
+import { PaginatedResponse, ResponseMeta } from "@metriport/shared";
+import { Request } from "express";
+import { z } from "zod";
+import {
+  Pagination,
+  PaginationFromItem,
+  PaginationItem,
+  PaginationToItem,
+} from "../command/pagination";
+
+export const defaultItemsPerPage = 50;
+export const maxItemsPerPage = 500;
+
+export const queryMetaSchema = z.intersection(
+  z.union(
+    [
+      z.object({
+        fromItem: z.string().optional(),
+        toItem: z.never().optional(),
+      }),
+      z.object({
+        fromItem: z.never().optional(),
+        toItem: z.string().optional(),
+      }),
+    ],
+    { errorMap: () => ({ message: "Either fromItem or toItem can be provided, but not both" }) }
+  ),
+  z.object({
+    count: numericValue.optional().refine(count => (count ? count <= maxItemsPerPage : true), {
+      message: `Count cannot be greater than ${maxItemsPerPage}`,
+    }),
+  })
+);
+export type HttpMeta = z.infer<typeof queryMetaSchema>;
+
+export function getRequestMeta(req: Request): Pagination {
+  const parsed = queryMetaSchema.parse(req.query);
+  return {
+    ...parsed,
+    ...(parsed.count ? { count: Number(parsed.count) } : { count: defaultItemsPerPage }),
+  };
+}
+
+/**
+ * Function to paginate a list of items.
+ *
+ * @param req - The HTTP request object.
+ * @param getItems - A function that takes pagination settings and returns a list of items for a given page.
+ * @returns An object containing the pagination metadata and the current page's items.
+ */
+export async function paginated<T extends { id: string }>(
+  req: Request,
+  getItems: (pagination: Pagination) => Promise<T[]>
+): Promise<PaginatedResponse<T, "items">> {
+  const requestMeta = getRequestMeta(req);
+
+  const { prevPageItem, nextPageItem, currPageItems } = await getPaginationItems(
+    requestMeta,
+    getItems
+  );
+
+  const responseMeta: ResponseMeta = {
+    ...(prevPageItem ? { prevPage: getPrevPageUrl(req, prevPageItem) } : {}),
+    ...(nextPageItem ? { nextPage: getNextPageUrl(req, nextPageItem) } : {}),
+    itemsOnPage: currPageItems.length,
+  };
+  return { meta: responseMeta, items: currPageItems };
+}
+
+async function getPaginationItems<T extends { id: string }>(
+  requestMeta: Pagination,
+  getItems: (filterAndPagination: Pagination) => Promise<T[]>
+): Promise<{
+  prevPageItem: string | undefined;
+  currPageItems: T[];
+  nextPageItem: string | undefined;
+}> {
+  const itemsPerPage = requestMeta.count;
+
+  // return the items for the current page + one more to determine if there is a next page
+  const itemsWithExtraOne = await getItems({
+    ...requestMeta,
+    count: itemsPerPage + 1,
+  });
+  if (itemsWithExtraOne.length < 1) {
+    return { prevPageItem: undefined, nextPageItem: undefined, currPageItems: [] };
```

Comment:
Couldn't there conceivably have been a previous page?

Reply:
We only have a page when there are items on that page. If there's no item, it's safe to assume there's no item at all - and we wouldn't be able to search items for the previous page if there's no item on the page.